Add OIDC backend configuration schema and validation

uwwint opened the pull request:
Add schema validation for oidc_backends_config.xml. 
The base mechanism can be used for all XML files.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

uwwint committed:
add schema validation to oidc_backends_config. This mechanism can be used for all XML files.

Mention accepted_audiences in oidc_backends_config.xml.sample

uwwint committed:
Use Union and lower_case param

uwwint committed:
Fix f string not being used correctly in exception logging.

uwwint committed:
Fix linting issues

uwwint committed:
Fix formatting issues

uwwint committed:
fix final formatting issue (finally make format doesn't die on me anymore)

mvdbeek commented on lib/galaxy/util/__init__.py:
```diff
@@ -294,15 +294,25 @@ def unique_id(KEY_SIZE=128):
+
         tree = etree.parse(str(fname), parser=parser)
+        if schema:
+            schema.assertValid(tree)
```
Can you do this at the end so the remaining processing occurs, and lxml is not a hard requirements-, so this should only be attempted if `LXML_AVAILABLE`

uwwint replied:
Should all be resolved now.

mvdbeek reviewed (commented):
It's a good idea to have schema for our config options, however I think we discussed removing all those little config files and just defining the config inline in galaxy.yml. In the meantime config/schemas is not the right location for this, I'd sugest lib/galaxy/authnz/xsd to follow the pattern used for other schemas.

uwwint committed:
move location of XSD to lib/galaxy/authnz/xsd
change processing order for validation to occur after pre-processing the XML
ensure validation is only done if lxml is available
raise validation error, such that issue is immediately visible to the user.

mvdbeek committed:
Use galaxy.util.resources.files

nsoranzo commented on lib/galaxy/util/__init__.py:
```diff
@@ -294,13 +294,22 @@ def unique_id(KEY_SIZE=128):
 
-def parse_xml(fname: StrPath, strip_whitespace=True, remove_comments=True) -> ElementTree:
+def parse_xml(
+    fname: StrPath, schemafname: Union[StrPath, None] = None, strip_whitespace=True, remove_comments=True
```
This could break the use of this method by third-party apps (via https://pypi.org/project/galaxy-util/ ), it would be safer to have `schemafname` be added as the last parameter).

mvdbeek replied:
Do you want to open a followup or should I do it ?

nsoranzo replied:
I can do it later.

nsoranzo commented on lib/galaxy/util/__init__.py:
```diff
@@ -318,6 +329,9 @@ def parse_xml(fname: StrPath, strip_whitespace=True, remove_comments=True) -> El
     except etree.ParseError:
         log.exception("Error parsing file %s", fname)
         raise
+    except etree.DocumentInvalid as e:
+        log.exception(f"Validation of file %s failed with error {e}" % fname)
```
This should have been:
```python
    except etree.DocumentInvalid:
        log.exception("Validation of file %s failed", fname)
```
